packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -477,35 +395,37 @@ contract ControllerHelper is UniswapControllerHelper, AaveControllerHelper, IERC
         uint8 _callSource,
         bytes memory _calldata
     ) internal override {
-        if (CALLBACK_SOURCE(_callSource) == CALLBACK_SOURCE.FLASHLOAN_W_MINT_DEPOSIT_NFT) {
-            FlashloanWMintDepositNftParams memory data = abi.decode(_calldata, (FlashloanWMintDepositNftParams));
+        if (
+            ControllerHelperDataType.CALLBACK_SOURCE(_callSource) ==
+            ControllerHelperDataType.CALLBACK_SOURCE.FLASHLOAN_W_MINT_DEPOSIT_NFT
+        ) {
+            ControllerHelperDataType.FlashloanWMintDepositNftParams memory data = abi.decode(
+                _calldata,
+                (ControllerHelperDataType.FlashloanWMintDepositNftParams)
+            );
 
             // convert flashloaned WETH to ETH
             IWETH9(weth).withdraw(_amount);
 
-            uint256 vaultId = IController(controller).mintWPowerPerpAmount{value: data.collateralToDeposit}(
-                data.vaultId,
-                data.wPowerPerpAmount,
-                0
-            );
-
-            // LP data.wPowerPerpAmount & data.collateralToLp in Uni v3
-            uint256 amount0Desired = isWethToken0 ? data.collateralToLp : data.wPowerPerpAmount;
-            uint256 amount1Desired = isWethToken0 ? data.wPowerPerpAmount : data.collateralToLp;
-            uint256 uniTokenId = _lpWPowerPerpPool(
-                address(this),
-                data.collateralToLp,
-                amount0Desired,
-                amount1Desired,
-                data.lpAmount0Min,
-                data.lpAmount1Min,
-                block.timestamp,
-                data.lpLowerTick,
-                data.lpUpperTick
+            (uint256 vaultId, uint256 uniTokenId) = ControllerHelperUtil.mintAndLp(
+                controller,
+                nonfungiblePositionManager,
+                wPowerPerp,
+                wPowerPerpPool,
+                ControllerHelperDataType.MintAndLpParams({
+                    recipient: address(this),
+                    vaultId: data.vaultId,
+                    wPowerPerpAmount: data.wPowerPerpAmount,
+                    collateralToDeposit: data.collateralToDeposit,
```

Comment:
doesn't this need to be data.collateralToDeposit.add(data.collateralToFlashloan) or .add(_amount), otherwise we don't deposit the _amount in the vault.

Reply:
This can work also if collateralToDeposit always have the flashloaned amount, but will change it